pkg/template/controller/templateinstance_controller.go

```diff
@@ -0,0 +1,174 @@
+import (
+	"net/http"
+
+	"github.com/golang/glog"
+	"github.com/openshift/origin/pkg/api/latest"
+	authclient "github.com/openshift/origin/pkg/auth/client"
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/config/cmd"
+	templateapi "github.com/openshift/origin/pkg/template/api"
+	templateapiv1 "github.com/openshift/origin/pkg/template/api/v1"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/meta/metatypes"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/auth/user"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type TemplateInstanceController struct {
+	restconfig restclient.Config
+	oc         *client.Client
+	controller *cache.Controller
+}
+
+func NewTemplateInstanceController(restconfig restclient.Config, oc *client.Client) *TemplateInstanceController {
+	c := TemplateInstanceController{restconfig: restconfig, oc: oc}
+
+	_, c.controller = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return c.oc.TemplateInstances(kapi.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return c.oc.TemplateInstances(kapi.NamespaceAll).Watch(options)
+			},
+		},
+		&templateapi.TemplateInstance{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.handle(obj.(*templateapi.TemplateInstance))
+			},
+			UpdateFunc: func(_, obj interface{}) {
+				c.handle(obj.(*templateapi.TemplateInstance))
+			},
+			DeleteFunc: func(obj interface{}) {
+			},
+		},
+	)
+
+	return &c
+}
+
+func (c *TemplateInstanceController) Run(stop <-chan struct{}) {
+	c.controller.Run(stop)
+}
+
+func (c *TemplateInstanceController) handle(templateInstance *templateapi.TemplateInstance) error {
+	for _, condition := range templateInstance.Status.Conditions {
+		if condition.Type == templateapi.TemplateInstanceCreated && condition.Status == kapi.ConditionTrue ||
+			condition.Type == templateapi.TemplateInstanceFailed && condition.Status == kapi.ConditionTrue {
+			return nil
+		}
+	}
+
+	err := c.provision(templateInstance)
+	if err == nil {
+		templateInstance.Status.Conditions = []templateapi.TemplateInstanceCondition{
+			{
+				Type:               templateapi.TemplateInstanceCreated,
+				Status:             kapi.ConditionTrue,
+				LastTransitionTime: unversioned.Now(),
+				Reason:             "Created",
+			},
+		}
+
+	} else {
+		templateInstance.Status.Conditions = []templateapi.TemplateInstanceCondition{
+			{
+				Type:               templateapi.TemplateInstanceFailed,
+				Status:             kapi.ConditionTrue,
+				LastTransitionTime: unversioned.Now(),
+				Reason:             "Failed",
```

Comment:
Two different conditions with inverse state is pretty weird.  Generally this would all be under the Created condition, or you'd create a "Provisioned" condition with Unknown as the initial state and true or false as the terminal state.  I guess the question is - is this controller *always* ensuring the template is provisioned (in which case there are no terminal states, and you must retry always), or is it a one shot thing.  If this was an operator, it'd be retry always.